lib/api/core/models/security/profile.js

```diff
@@ -55,46 +82,48 @@ Profile.prototype.validateDefinition = function () {
  *
  * @return {Promise}
  */
-Profile.prototype.getRights = function() {
+Profile.prototype.getRights = function(kuzzle) {
   var profileRights = {};
 
-  this.roles.forEach(role => {
-    Object.keys(role.controllers).forEach(controller => {
-      var restrictedTo = _.cloneDeep(role.restrictedTo);
-
-      Object.keys(role.controllers[controller].actions).forEach(action => {
-        var actionRights = role.controllers[controller].actions[action];
-
-        if (restrictedTo === undefined || restrictedTo.length === 0) {
-          restrictedTo = [{index: '*', collections: ['*']}];
-        }
-
-        restrictedTo.forEach(restriction => {
-          if (restriction.collections === undefined || restriction.collections.length === 0) {
-            restriction.collections = ['*'];
-          }
-
-          restriction.collections.forEach(collection => {
-            var
-              rightsObject = {},
-              rightsItem = {
-                controller: controller,
-                action: action,
-                index: restriction.index,
-                collection: collection
-              },
-              rightsKey = md5(JSON.stringify(rightsItem));
-
-            rightsItem.value = actionRights;
-            rightsObject[rightsKey] = rightsItem;
-            profileRights = _.assignWith(profileRights, rightsObject, mergeRights);
+  return this.getRoles(kuzzle)
+    .then(roles => {
+      roles.forEach(role => {
+        Object.keys(role.controllers).forEach(controller => {
+          var restrictedTo = _.cloneDeep(role.restrictedTo);
```

Comment:
I gave a long hard look to this line, and after some time I came to the conclusion that we don't need to perform this `_.cloneDeep` more than once per role, instead of once per role controller like it's done here.

Should be in the first `roles.forEach()` loop instead.